From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lvyufeng <lvyufeng@users.noreply.github.com>
Date: Thu, 4 Jun 2026 17:12:00 +0000
Subject: [PATCH] gpio: litex: add inout tristate GPIO support

LiteX GPIOTristate exposes oe/in/out registers before optional IRQ
registers. Add a litex,direction = "inout" mode so GPIO consumers can
switch a line between high-Z input and driven output. This is needed by
ps2-gpio/open-drain consumers on the Nexys4 DDR PS/2 clock/data lines.

Also provide irq_enable/irq_disable callbacks in addition to mask/unmask so
Linux's generic IRQ disable path updates gpiolib's enabled-IRQ state before a
consumer temporarily changes a GPIO line direction.
---
diff --git a/drivers/gpio/gpio-litex.c b/drivers/gpio/gpio-litex.c
--- a/drivers/gpio/gpio-litex.c
+++ b/drivers/gpio/gpio-litex.c
@@ -36,6 +36,15 @@
 #define LITEX_GPIO_PENDING_OFFSET 0x10
 #define LITEX_GPIO_ENABLE_OFFSET 0x14
 
+#define LITEX_GPIO_DIR_INOUT 2
+#define LITEX_GPIO_TRISTATE_OE_OFFSET 0x0
+#define LITEX_GPIO_TRISTATE_IN_OFFSET 0x4
+#define LITEX_GPIO_TRISTATE_OUT_OFFSET 0x8
+#define LITEX_GPIO_TRISTATE_MODE_OFFSET 0xc
+#define LITEX_GPIO_TRISTATE_EDGE_OFFSET 0x10
+#define LITEX_GPIO_TRISTATE_PENDING_OFFSET 0x18
+#define LITEX_GPIO_TRISTATE_ENABLE_OFFSET 0x1c
+
 struct litex_gpio {
 	void __iomem *membase;
 	int port_direction;
@@ -58,6 +67,47 @@
 	_litex_set_reg(gpio_s->membase + reg_offset, gpio_s->reg_span, value);
 }
 
+static inline bool litex_gpio_is_inout(struct litex_gpio *gpio_s)
+{
+	return gpio_s->port_direction == LITEX_GPIO_DIR_INOUT;
+}
+
+static inline int litex_gpio_input_offset(struct litex_gpio *gpio_s)
+{
+	return litex_gpio_is_inout(gpio_s) ?
+		LITEX_GPIO_TRISTATE_IN_OFFSET : LITEX_GPIO_VALUE_OFFSET;
+}
+
+static inline int litex_gpio_output_offset(struct litex_gpio *gpio_s)
+{
+	return litex_gpio_is_inout(gpio_s) ?
+		LITEX_GPIO_TRISTATE_OUT_OFFSET : LITEX_GPIO_VALUE_OFFSET;
+}
+
+static inline int litex_gpio_mode_offset(struct litex_gpio *gpio_s)
+{
+	return litex_gpio_is_inout(gpio_s) ?
+		LITEX_GPIO_TRISTATE_MODE_OFFSET : LITEX_GPIO_MODE_OFFSET;
+}
+
+static inline int litex_gpio_edge_offset(struct litex_gpio *gpio_s)
+{
+	return litex_gpio_is_inout(gpio_s) ?
+		LITEX_GPIO_TRISTATE_EDGE_OFFSET : LITEX_GPIO_EDGE_OFFSET;
+}
+
+static inline int litex_gpio_pending_offset(struct litex_gpio *gpio_s)
+{
+	return litex_gpio_is_inout(gpio_s) ?
+		LITEX_GPIO_TRISTATE_PENDING_OFFSET : LITEX_GPIO_PENDING_OFFSET;
+}
+
+static inline int litex_gpio_enable_offset(struct litex_gpio *gpio_s)
+{
+	return litex_gpio_is_inout(gpio_s) ?
+		LITEX_GPIO_TRISTATE_ENABLE_OFFSET : LITEX_GPIO_ENABLE_OFFSET;
+}
+
 /* API functions */
 
 static int litex_gpio_get_value(struct gpio_chip *chip, unsigned int offset)
@@ -68,7 +118,7 @@
 	if (offset >= chip->ngpio)
 		return -EINVAL;
 
-	regv = _litex_get_reg(gpio_s->membase, gpio_s->reg_span);
+	regv = litex_gpio_get_reg(gpio_s, litex_gpio_input_offset(gpio_s));
 	return !!(regv & BIT(offset));
 }
 
@@ -81,7 +131,7 @@
 	if (*mask >= (1 << chip->ngpio))
 		return -EINVAL;
 
-	regv = _litex_get_reg(gpio_s->membase, gpio_s->reg_span);
+	regv = litex_gpio_get_reg(gpio_s, litex_gpio_input_offset(gpio_s));
 	*bits = (regv & *mask);
 	return 0;
 }
@@ -91,13 +141,15 @@
 {
 	struct litex_gpio *gpio_s = gpiochip_get_data(chip);
 	u32 regv, new_regv;
+	int reg_offset;
 
 	if (offset >= chip->ngpio)
 		return;
 
-	regv = _litex_get_reg(gpio_s->membase, gpio_s->reg_span);
+	reg_offset = litex_gpio_output_offset(gpio_s);
+	regv = litex_gpio_get_reg(gpio_s, reg_offset);
 	new_regv = (regv & ~BIT(offset)) | (!!val << offset);
-	_litex_set_reg(gpio_s->membase, gpio_s->reg_span, new_regv);
+	litex_gpio_set_reg(gpio_s, reg_offset, new_regv);
 }
 
 static void litex_gpio_set_multiple(struct gpio_chip *chip, unsigned long *mask,
@@ -105,18 +157,30 @@
 {
 	struct litex_gpio *gpio_s = gpiochip_get_data(chip);
 	u32 regv, new_regv;
+	int reg_offset;
 
 	if (*mask >= (1 << chip->ngpio))
 		return;
 
-	regv = _litex_get_reg(gpio_s->membase, gpio_s->reg_span);
+	reg_offset = litex_gpio_output_offset(gpio_s);
+	regv = litex_gpio_get_reg(gpio_s, reg_offset);
 	new_regv = (regv & ~(*mask)) | (*bits);
-	_litex_set_reg(gpio_s->membase, gpio_s->reg_span, new_regv);
+	litex_gpio_set_reg(gpio_s, reg_offset, new_regv);
 }
 
 static int litex_gpio_get_direction(struct gpio_chip *chip, unsigned int offset)
 {
 	struct litex_gpio *gpio_s = gpiochip_get_data(chip);
+	u32 oe;
+
+	if (offset >= chip->ngpio)
+		return -EINVAL;
+
+	if (litex_gpio_is_inout(gpio_s)) {
+		oe = litex_gpio_get_reg(gpio_s, LITEX_GPIO_TRISTATE_OE_OFFSET);
+		return (oe & BIT(offset)) ?
+			GPIO_LINE_DIRECTION_OUT : GPIO_LINE_DIRECTION_IN;
+	}
 
 	return gpio_s->port_direction;
 }
@@ -125,6 +189,20 @@
 				      unsigned int offset)
 {
 	struct litex_gpio *gpio_s = gpiochip_get_data(chip);
+	unsigned long flags;
+	u32 oe;
+
+	if (offset >= chip->ngpio)
+		return -EINVAL;
+
+	if (litex_gpio_is_inout(gpio_s)) {
+		spin_lock_irqsave(&gpio_s->gpio_lock, flags);
+		oe = litex_gpio_get_reg(gpio_s, LITEX_GPIO_TRISTATE_OE_OFFSET);
+		oe &= ~BIT(offset);
+		litex_gpio_set_reg(gpio_s, LITEX_GPIO_TRISTATE_OE_OFFSET, oe);
+		spin_unlock_irqrestore(&gpio_s->gpio_lock, flags);
+		return 0;
+	}
 
 	if (gpio_s->port_direction != GPIOF_DIR_IN)
 		return -ENOTSUPP;
@@ -136,6 +214,23 @@
 				       unsigned int offset, int value)
 {
 	struct litex_gpio *gpio_s = gpiochip_get_data(chip);
+	unsigned long flags;
+	u32 regv, oe;
+
+	if (offset >= chip->ngpio)
+		return -EINVAL;
+
+	if (litex_gpio_is_inout(gpio_s)) {
+		spin_lock_irqsave(&gpio_s->gpio_lock, flags);
+		regv = litex_gpio_get_reg(gpio_s, LITEX_GPIO_TRISTATE_OUT_OFFSET);
+		regv = (regv & ~BIT(offset)) | (!!value << offset);
+		litex_gpio_set_reg(gpio_s, LITEX_GPIO_TRISTATE_OUT_OFFSET, regv);
+		oe = litex_gpio_get_reg(gpio_s, LITEX_GPIO_TRISTATE_OE_OFFSET);
+		oe |= BIT(offset);
+		litex_gpio_set_reg(gpio_s, LITEX_GPIO_TRISTATE_OE_OFFSET, oe);
+		spin_unlock_irqrestore(&gpio_s->gpio_lock, flags);
+		return 0;
+	}
 
 	if (gpio_s->port_direction != GPIOF_DIR_OUT)
 		return -ENOTSUPP;
@@ -157,10 +252,10 @@
 	spin_lock_irqsave(&gpio_s->gpio_lock, flags);
 
 	/* Clear any sticky pending interrupts */
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_PENDING_OFFSET, bit);
-	enable = litex_gpio_get_reg(gpio_s, LITEX_GPIO_ENABLE_OFFSET);
+	litex_gpio_set_reg(gpio_s, litex_gpio_pending_offset(gpio_s), bit);
+	enable = litex_gpio_get_reg(gpio_s, litex_gpio_enable_offset(gpio_s));
 	enable |= bit;
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_ENABLE_OFFSET, enable);
+	litex_gpio_set_reg(gpio_s, litex_gpio_enable_offset(gpio_s), enable);
 
 	spin_unlock_irqrestore(&gpio_s->gpio_lock, flags);
 }
@@ -176,9 +271,9 @@
 
 	spin_lock_irqsave(&gpio_s->gpio_lock, flags);
 
-	enable = litex_gpio_get_reg(gpio_s, LITEX_GPIO_ENABLE_OFFSET);
+	enable = litex_gpio_get_reg(gpio_s, litex_gpio_enable_offset(gpio_s));
 	enable &= ~bit;
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_ENABLE_OFFSET, enable);
+	litex_gpio_set_reg(gpio_s, litex_gpio_enable_offset(gpio_s), enable);
 
 	spin_unlock_irqrestore(&gpio_s->gpio_lock, flags);
 
@@ -196,8 +291,8 @@
 
 	spin_lock_irqsave(&gpio_s->gpio_lock, flags);
 
-	mode = litex_gpio_get_reg(gpio_s, LITEX_GPIO_MODE_OFFSET);
-	edge = litex_gpio_get_reg(gpio_s, LITEX_GPIO_EDGE_OFFSET);
+	mode = litex_gpio_get_reg(gpio_s, litex_gpio_mode_offset(gpio_s));
+	edge = litex_gpio_get_reg(gpio_s, litex_gpio_edge_offset(gpio_s));
 
 	switch (type & IRQ_TYPE_SENSE_MASK) {
 		case IRQ_TYPE_NONE:
@@ -220,8 +315,8 @@
 		default:
 			return -EINVAL;
 	}
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_MODE_OFFSET, mode);
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_EDGE_OFFSET, edge);
+	litex_gpio_set_reg(gpio_s, litex_gpio_mode_offset(gpio_s), mode);
+	litex_gpio_set_reg(gpio_s, litex_gpio_edge_offset(gpio_s), edge);
 
 	spin_unlock_irqrestore(&gpio_s->gpio_lock, flags);
 
@@ -239,7 +334,7 @@
 	spin_lock_irqsave(&gpio_s->gpio_lock, flags);

 	/* Clear all pending interrupts */
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_PENDING_OFFSET, bit);
+	litex_gpio_set_reg(gpio_s, litex_gpio_pending_offset(gpio_s), bit);

 	spin_unlock_irqrestore(&gpio_s->gpio_lock, flags);

+@@ -252,6 +347,8 @@
+ static const struct irq_chip litex_irq_chip = {
+ 	.name			= "litex-gpio",
++	.irq_enable		= litex_gpio_irq_unmask,
++	.irq_disable		= litex_gpio_irq_mask,
+ 	.irq_unmask		= litex_gpio_irq_unmask,
+ 	.irq_mask		= litex_gpio_irq_mask,
+ 	.irq_set_type		= litex_gpio_irq_set_type,
@@ -291,8 +386,8 @@
 
 	chained_irq_enter(ichip, desc);
 
-	enabled = litex_gpio_get_reg(gpio_s, LITEX_GPIO_ENABLE_OFFSET);
-	pending = litex_gpio_get_reg(gpio_s, LITEX_GPIO_PENDING_OFFSET);
+	enabled = litex_gpio_get_reg(gpio_s, litex_gpio_enable_offset(gpio_s));
+	pending = litex_gpio_get_reg(gpio_s, litex_gpio_pending_offset(gpio_s));
 	interrupts_to_handle = pending & enabled;
 
 	for_each_set_bit(pin, &interrupts_to_handle, GPIO_PINS_MAX) {
@@ -330,7 +425,7 @@
 	gpio_s->irq_number = platform_get_irq(pdev, 0);
 
 	/* Disable all GPIO interrupts before enabling parent interrupts */
-	litex_gpio_set_reg(gpio_s, LITEX_GPIO_ENABLE_OFFSET, 0);
+	litex_gpio_set_reg(gpio_s, litex_gpio_enable_offset(gpio_s), 0);
 
 	gichip = &gpio_s->chip.irq;
 	gpio_irq_chip_set_chip(gichip, &litex_irq_chip);
@@ -397,6 +492,8 @@
 		gpio_s->port_direction = GPIOF_DIR_IN;
 	else if (!strcmp(dt_direction, "out"))
 		gpio_s->port_direction = GPIOF_DIR_OUT;
+	else if (!strcmp(dt_direction, "inout"))
+		gpio_s->port_direction = LITEX_GPIO_DIR_INOUT;
 	else
 		return -ENODEV;
 
@@ -419,7 +516,8 @@
 	gpio_s->reg_span = (dt_ngpio + LITEX_SUBREG_SIZE_BIT - 1) /
 			   LITEX_SUBREG_SIZE_BIT;
 
-	if (gpio_s->port_direction == GPIOF_DIR_IN) {
+	if (gpio_s->port_direction == GPIOF_DIR_IN ||
+	    litex_gpio_is_inout(gpio_s)) {
 		ret_i = litex_gpio_init_irq(pdev, gpio_s);
 		if (ret_i < 0)
 			return ret_i;
